src/System.Private.CoreLib/shared/System/Diagnostics/DebugProvider.cs

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Do not remove this, it is needed to retain calls to these conditional methods in release builds
+#define DEBUG
+
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// Provides default implementation for Write and ShowDialog methods in Debug class.
+    /// </summary>
+    public partial class DebugProvider
+    {
+        public virtual void Write(string message)
+        {
+            lock (s_lock)
+            {
+                if (message == null)
+                {
+                    s_WriteCore(string.Empty);
```

Comment:
The DebugTests want to be able to swap out the behavior of s_WriteCore here without overriding the indent logic in this Write method.

DebugTests are trying to validate the Write and Assert methods in Debug which both call DebugProvider.Write here, will format the `message` input string properly. 
Because of this I can't be using a derived DebugProvider class in DebugTests to also override Write (the same it's done for ShowDialog [here](https://github.com/dotnet/corefx/pull/32846/files#diff-807b66f6424dc6bc60cc2ebec979ae53L230)).

Reply:
As I have mentioned offline, the tests that are doing this swapping are not very useful anymore since you should be able to exercise the same code paths via the public Trace Listener now.

Reply:
In this PR I did not consider fixing existing indentation bugs and inconsistencies in Debug class. (refer to [this diff](https://github.com/maryamariyan/corefx/pull/30/files)).

In this PR I provide the DebugProvider API and a mechanism to Wire Debug with Trace Listeners without changing any indentation behavior in Debug.

As you see in DebugProvider.cs, the Write method has both logic of writing to debugger and indentation coupled inside of it. The only way, as is, to be able to test indentation while mocking writing to debugger is to keep the delegate.

To keep track of existing Debug indentation behavior and the discrepancies until the fix is in place I think there is value in keeping DebugTests and hence the s_WriteCore delegate.
In my follow up PR I will remove s_WriteCore delegate as well as fix the indentation problems.

Please let me know what you think.
I could either keep the the scope of this PR limited to just adding DebugProvider or increase the scope to also fix indentation in which case the delegate would also be remove as part of this PR.

Reply:
ok